# Cystic Fibrosis Antibiotic Resistance Gene Dynamics
Exploring the antibiotic resistance dynamics in infants with CF as well as non-CF infants using shotgun metagenomic sequencing data. We have longitudinal shotgun metagenomic sequencing data from 40 infants in the first three years of life, for a total of 190 samples. These samples have been explored by our lab previously and the results are currently on [BioRxiv](https://www.biorxiv.org/content/10.1101/2023.05.02.539134v1). 

Now, we will be expanding our findings by looking at **antibiotic resistance gene (ARG)** abundance in these samples. We will be doing so by using [RGI v6.0.2](https://github.com/arpcard/rgi), which is a software developed using the [Comprehensive Antibiotic Resistance Database (CARD) v3.2.7](https://card.mcmaster.ca/) to quantify ARGs. The methodology used in this project was inspired by [Li et al. 2021](https://www.sciencedirect.com/science/article/pii/S1931312821001451). 

I'll also be using several datasets to compare the DartCF cohort to, including:
[Li et al. 2021 (Original ARG paper)](https://www.sciencedirect.com/science/article/pii/S1931312821001451)
[Stewart et. al 2018 (TEDDY paper)](https://www.nature.com/articles/s41586-018-0617-x)
[Yassour et. al 2016 (ARG focused paper, DIABIMMUNE cohort)](https://www.science.org/doi/10.1126/scitranslmed.aad0917)
[Vatanen et. al 2016 (DIABIMMUNE cohort)](https://www.sciencedirect.com/science/article/pii/S0092867416303981)
[Kostic et. al 2015 (DIABIMMUNE cohort)](https://www.cell.com/cell-host-microbe/pdf/S1931-3128(15)00021-9.pdf)
[Hayden et. al 2020 (CF cohort)](https://www.nature.com/articles/s41591-019-0714-x)
